WGSL: describe memory layout rules for buffers

dneto0 opened the pull request:
This builds on #1214 (describe resource interface of a shader)  


  WGSL: describe memory layout rules for buffers
    
    - Add notations for floor, ceiling, roundUp
    - define 'stride' and 'offset' attributes
    - describe byte-by-byte layout of values, based on storage class, and
    stride and offset
    - describe constraints on stride and offset, yielding 'uniform buffer
    layout' and 'storage buffer layout'

dj2 commented on wgsl/index.bs:
```diff
@@ -581,28 +615,300 @@ storage_class
+not support the `scalarBlockLayout` feature.
+Therefore, a data value must not be placed in the padding at the end of a structure or matrix,
+nor in the padding at the last element of an array.
+Counting the such padding as part of the allocation extent allows [SHORTNAME] to capture this constraint.
```
nit: `Counting the padding` or `Counting such padding`

dneto0 replied:
Fixed in 9603cf

dj2 commented on wgsl/index.bs:
```diff
@@ -3165,6 +3469,25 @@ where compatibility is defined by the following table.
 </table>
 
+If |B| is a [=uniform buffer=] variable in a resource interface,
+and |WB| is the WebGPU GPUBuffer bound to |B|, then:
```
Can we link `WebGPU GPUBuffer` back to the spec?

dneto0 replied:
Fixed in 9603cf

dneto0 commented on wgsl/index.bs:
```diff
@@ -581,28 +615,300 @@ storage_class
 in a pipeline, or between different pipelines.
 
-Because the data are shared without reformatting or translation,
+Because buffer data are shared without reformatting or translation,
```
data is plural.

dj2 replied:
It was the `buffer` part I cared about, not the `is` part, heh.

dneto0 commented on wgsl/index.bs:
```diff
@@ -581,28 +615,300 @@ storage_class
 The memory layout of a type is significant only when referring to a value in those storage classes.
-This affects evaluation of a variable in one of those storage classes, or of a pointer into one of those storage classes.
+This affects evaluation of a variable in one of those storage classes,
+or of a pointer expression pointing into one of those storage classes.
```
Replacing those last two sentences with:

> The memory layout of a type is significant only when evaluating an expression with:
> *  a variable in the uniform or storage storage class, or
> * a pointer into the uniform or storage storage class.

dneto0 replied:
PTAL at 9603cf

dneto0 commented on wgsl/index.bs:
```diff
@@ -581,28 +615,300 @@ storage_class
+   * |V|.x is placed at byte offset |k|
+   * |V|.y is placed at byte offset |k|+4
+   * If |N| &ge; 3, then |V|.z is placed at byte offset |k|+8
+   * If |N| &ge; 4, then |V|.w is placed at byte offset |k|+12
```
I could.  But that makes everything harder to check and more confusing.

Whomever adds the f16 feature will have to modify these tables.

dneto0 commented:
Rebased, and added 9603cf to reply to most of @dj2's feedback.

dneto0 commented on wgsl/index.bs:
```diff
@@ -469,12 +490,25 @@ The following types are <dfn dfn noexport>host-shareable</dfn>:
+  <tr><td><dfn noexport>`stride`</dfn>
+      <td>positive i32 literal
+      <td>Applied to an array type.<br>
+         The number of bytes between successive elements of the array,
```
Fixed in 9603cf

dneto0 commented:
The parent PR was updated, so I rebased again.

grorg commented:
Discussed at [2020-11-17 meeting](https://docs.google.com/document/d/1JMqCbDXyuEvjH8GalT8_bEJ9vfejMo4GMIsAVaXnKvI/).

dneto0 commented:
From the meeting:
- add a non-normative statement about how std140 and std430 from OpenGL (and GLSL) map to these concepts
- add non-normative statement about what Vulkan concepts map to these

kvark commented on wgsl/index.bs:
```diff
@@ -600,28 +642,334 @@ storage_class
+See [[#resource-layout-compatibility]].
+
+Compared to OpenGL:
+* OpenGL `std140` layout satisfies [SHORTNAME] [=uniform buffer layout=].
```
is it "satisfies", or is it "exactly matches"?

dneto0 replied:
Satisfies.  WGSL uniform buffer layout can give strides and offsets that are multiples of the ones specified by std140.

Example:

      uint arr[3];   // will have stride of 16 in std140. 

But 

      [[stride(32)]] array<u32,3>; // This is perfectly fine to use inside a uniform buffer in WGSL

kvark replied:
Can we instead say something like
> For any type |S|, the tightest offset and stride assignments would exactly match both [SHORTNAME] [=uniform buffer layout=] and `std140`

kvark replied:
This is unresolved, right?

dneto0 replied:
How about:

* For any type |S|, OpenGL std140 layout is the same as using the tightest offset and stride assignments in [SHORTNAME] [=uniform buffer layout=].

dneto0 replied:
Or remove the quantification by |S|:

*  For any type, OpenGL `std140` layout is the same as using the tightest offset and stride assignments in WGSL [=uniform buffer layout=].

kvark commented on wgsl/index.bs:
```diff
@@ -600,28 +642,334 @@ storage_class
+Because buffer data are shared without reformatting or translation,
 buffer producers and consumers must agree on the <dfn noexport>memory layout</dfn>,
-which is the description of how the buffer's bytes are organized into typed WGSL values.
+which is the description of how the bytes in a buffer are organized into typed WGSL values.
```
should this use `[SHORTNAME]`?

dneto0 replied:
will fix.

kvark commented on wgsl/index.bs:
```diff
@@ -600,28 +642,334 @@ storage_class
+    * Vulkan allows a vector to be aligned to the size of its scalar component, but [SHORTNAME] requires a more constrained alignment.
+    * The Vulkan `scalarBlockLayout` and `uniformBufferStandardLayout` features do not apply to [SHORTNAME].
+    * The Vulkan concept of *scalar alignment* does not correspond to a concept in [SHORTNAME];
+* The Vulkan *base alignment* for a type |S| corresponds to the [SHORTNAME] alignment requirement for |S| in the `storage` storage class: |Align|(|S|,`storage`).
```
isn't it the other way around? I.e. "base alignment" == "uniform", and "extended alignment" == "storage"?

dneto0 replied:
You'd think it would be that way.  But it's not.  See [Offset and Stride Alignment](https://www.khronos.org/registry/vulkan/specs/1.0-extensions/html/vkspec.html#interfaces-resources-layout)

> The *extended alignment* of the type of an OpTypeStruct member is similarly defined as follows:

...

> An array or structure type has an extended alignment equal to the largest extended alignment of any of its members, rounded up to a multiple of 16.

To me, that's one of the confusing things about how it's expressed in Vulkan. Especially because you need an *extension* to use the *base* alignments on uniform buffer, but you *don't* use an extension to use the *extended* aligments in a uniform buffer.  (Ok, I purposely put the most confusing aspects together here, but the point holds)

kvark replied:
I see, was reading that part wrong. Thanks for clarification!

kvark commented on wgsl/index.bs:
```diff
@@ -600,28 +642,334 @@ storage_class
+
+This section describes how the internals of a value are placed in the byte locations
+of a buffer, given an assumed placement of the overall value.
+These layouts depend on the value's type, the storage class of the buffer,
```
should there be a separate sentence clarifying the behavior when the same type is used for both uniform and storage buffers?

dneto0 replied:
Will add that.

kvark commented on wgsl/index.bs:
```diff
@@ -600,28 +642,334 @@ storage_class
+A value |V| of type [=f32=] is represented in IEEE 754 binary32 format.
+It has one sign bit, 8 exponent bits, and 23 fraction bits.
+When |V| is placed at byte offset |k| of host-shared buffer, then:
+   * Byte |k| contains bits 0 through 7 of the fraction.
```
I wonder if this can be nicer visually encoded, as a follow up. For a crazy idea, it could be an 1x32 table, where a bunch of columns are merged.

dneto0 replied:
Definitely. I was going to look at that as an editorial improvement.    I'd have to learn something about drawing diagrams.

dneto0 commented:
Hey @kvark please take a look.

kvark reviewed (commented):
This is some grand work, thank you for writing it down!

dneto0 commented on wgsl/index.bs:
```diff
@@ -600,28 +642,334 @@ storage_class
+            = |k| &times; <var ignore>Align</var>(|T|<sub>|i|</sub>,<var ignore>C</var>),
+            for some non-negative integer |k|
+        </p>
+    * For the `uniform` storage class, members are aligned to 16 byte boundaries:
```
Oof.  I think this is wrong.  Will remove it.

dneto0 commented on wgsl/index.bs:
```diff
@@ -480,6 +501,13 @@ The following kinds of values must be of IO-shareable type:
+the host and the GPU, or copied between host and GPU without format translation.
+When used for this purpose, the type must be additionally decorated with layout attributes
+as described in [[#memory-layouts]].
+In particular, any variable using the `uniform` or `storage` [=storage classes=] must have
```
This is here to provide context. Really it's a forward reference.

I will change to:

> We will see in [[#module-scope-variables]] that the [=store type=] of [=uniform buffer=] and [=storage buffer=]
variables must be host-shareable.

dneto0 reviewed (commented):
I think I've addressed all feedback from @kvark

dneto0 commented:
Resolved @kvark's issue with std140, std430 (I hope).

Then I rebased against main, and fixed up the cross-references for storage classes as mentioned in the new text.
